compiler/rustc_data_structures/src/vec_cache.rs

```diff
@@ -0,0 +1,266 @@
+        // in-bounds of buckets. See `from_index` for computation.
+        let bucket = unsafe { buckets.get_unchecked(self.bucket_idx) };
+        let ptr = bucket.load(Ordering::Acquire);
+        // Bucket is not yet initialized: then we obviously won't find this entry in that bucket.
+        if ptr.is_null() {
+            return None;
+        }
+        // SAFETY: Follows from preconditions on `buckets` and `self`.
+        let slot = unsafe { ptr.add(self.index_in_bucket) };
+
+        // SAFETY:
+        let index_and_lock =
+            unsafe { &*slot.byte_add(offset_of!(Slot<V>, index_and_lock)).cast::<AtomicU32>() };
+        let current = index_and_lock.load(Ordering::Acquire);
+        let index = match current {
+            0 => return None,
+            // Treat "initializing" as actually just not initialized at all.
+            // The only reason this is a separate state is that `complete` calls could race and
+            // we can't allow that, but from load perspective there's no difference.
+            1 => return None,
+            _ => current - 2,
+        };
+
+        // SAFETY:
+        // * slot is a valid pointer (buckets are always valid for the index we get).
+        // * value is initialized since we saw a >= 2 index above.
+        // * `V: Copy`, so safe to read.
+        let value = unsafe { slot.byte_add(offset_of!(Slot<V>, value)).cast::<V>().read() };
+        Some((value, index))
+    }
+
+    /// Returns true if this successfully put into the map.
+    #[inline]
+    fn put<V>(&self, buckets: &[AtomicPtr<Slot<V>>; 21], value: V, extra: u32) -> bool {
+        static LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());
+
+        // SAFETY: `bucket_idx` is ilog2(u32).saturating_sub(11), which is at most 21, i.e.,
+        // in-bounds of buckets.
+        let bucket = unsafe { buckets.get_unchecked(self.bucket_idx) };
+        let mut ptr = bucket.load(Ordering::Acquire);
+        let _allocator_guard;
+        if ptr.is_null() {
+            // If we load null, then acquire the global lock; this path is quite cold, so it's cheap
+            // to use a global lock.
+            _allocator_guard = LOCK.lock();
+            // And re-load the value.
+            ptr = bucket.load(Ordering::Acquire);
+        }
+
+        // OK, now under the allocator lock, if we're still null then it's definitely us that will
+        // initialize this bucket.
+        if ptr.is_null() {
+            let bucket_layout =
+                std::alloc::Layout::array::<Slot<V>>(self.entries as usize).unwrap();
+            // SAFETY: Always >0 entries in each bucket.
+            let allocated = unsafe { std::alloc::alloc_zeroed(bucket_layout).cast::<Slot<V>>() };
+            if allocated.is_null() {
+                std::alloc::handle_alloc_error(bucket_layout);
+            }
+            bucket.store(allocated, Ordering::Release);
+            ptr = allocated;
+        }
+        assert!(!ptr.is_null());
```

Comment:
vibeck: i feel like this should be part of the `if ptr.is_null()` from above. In case the check failed the first time, this branch will never get taken and even if it doesn't matter for perf, it does feel cleaner to me that way